### ISSUE TITLE
UrlResolutionTagHelper does not handle srcset attributes correctly

### ISSUE THREAD

**Issue by ctolkien** [aspnet/Mvc#3603]
UrlResolutionTagHelper [shorts out if the URL being parsed in does not start with "~/"](https://github.com/aspnet/Mvc/blob/52f4a831397b94d41675cabf8777b735a69e9c2f/src/Microsoft.AspNet.Mvc.Razor/TagHelpers/UrlResolutionTagHelper.cs#L203), however for `img` and `source` tags using the `srcset` attribute, the URL may not always be at the beginning. E.g:

```html
<img src="~/logo.png" srcset="~/logo.png 1x, ~/logo@2x.png 2x" />
```

Expected result:

```html
<img src="logo.png" srcset="logo.png 1x, logo@2x.png 2x" />
```

Actual Result:
```html
<img src="logo.png" srcset="logo.png 1x, ~/logo@2x.png 2x" />
```

**Comment by Eilon** [comment]
Moving this to backlog.

My recommendation would be to use `<img src="..." srcset="@Url.Content("~/img1.png") 1px @Url.Content("~/img2.png") 2px" ... />` as a workaround.

The actual fix for this is I think a bit more complex than the associated PR because it has to properly parse the full syntax as opposed to just splitting on commas. It has to split on commas, then also on the first space, and resolve only the URL portion (and not pass the optional specifier to the URL resolution logic).

**Comment by ctolkien** [comment]
@Eilon All good, I was scratching my own itch.

In the case where `encodeWebRoot` is true, the full app path just chops off `~/` and prefixes the application path, so that should be OK. However when it's false, the URL and specifier do get passed to `Url.Content` and I'm not sure on the full impact there. All the tests pass but I didn't dig any deeper.

Anyway, glad this feature is on the radar now that all modern browsers support `srcset`.

**Comment by Eilon** [comment]
Yeah it's a good feature to have, for sure. Thinking about it again, if you want to update the PR with the updated parsing logic then we could certainly re-evaluate. We're trying to focus on just code stability and not features so that's why I closed it earlier, but now that I think about it again, this is important from a standards compatibility perspective.

**Comment by ctolkien** [comment]
Hi @Eilon, dredging this one back up and am looking to re-implement, obviously a fair bit of cheese has been shuffled around since then.

> "Thinking about it again, if you want to update the PR with the updated parsing logic then we could certainly re-evaluate."

> "The actual fix for this is I think a bit more complex than the associated PR because it has to properly parse the full syntax as opposed to just splitting on commas. It has to split on commas, then also on the first space, and resolve only the URL portion (and not pass the optional specifier to the URL resolution logic)."

Digging into the URL resolution logic, I can't see any issue with passing more than the URL, as all it does is send back the application path and the URL with the leading `~` removed. Relevant bits:

https://github.com/aspnet/Mvc/blob/dev/src/Microsoft.AspNetCore.Mvc.Core/Routing/UrlHelper.cs#L197

Seems we can avoid doing any more in-depth parsing?

**Comment by Eilon** [comment]
@ctolkien unfortunately that just happens to be the current implementation. I've seen apps where the URL resolution logic is overridden to support more advanced scenarios (e.g. CDNs, other custom lookups, etc.), so I don't think it's safe to assume it's just simple logic with the `~` char...

**Comment by ctolkien** [comment]
Somewhat related, `asp-append-version="true"` also does not apply to `srcset` on `img` tags.
